src/dos/cdrom_win32.cpp

```diff
@@ -0,0 +1,229 @@
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#if defined(WIN32)
+
+#include "cdrom.h"
+
+#include <Windows.h>
+#include <Ntddcdrm.h>
+
+static char get_drive_letter(const char* path)
+{
+	const size_t len = strlen(path);
+	if (len < 2 || len > 3) {
+		return 0;
+	}
+	if (path[1] != ':') {
+		return 0;
+	}
+	const char drive_letter = toupper(path[0]);
+	if (drive_letter >= 'A' && drive_letter <= 'Z') {
+		return drive_letter;
+	}
+	return 0;
+}
+
+bool CDROM_Interface_Win32::IsOpen() const
+{
+	return cdrom_handle != INVALID_HANDLE_VALUE;
+}
+
+CDROM_Interface_Win32::~CDROM_Interface_Win32()
+{
+	if (IsOpen()) {
+		CloseHandle(cdrom_handle);
+	}
+	cdrom_handle = INVALID_HANDLE_VALUE;
+}
+
+bool CDROM_Interface_Win32::Open(const char drive_letter)
+{
+	const std::string device_path = std::string("\\\\.\\") + drive_letter + ':';
+
+	HANDLE device = CreateFileA(device_path.c_str(), GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE, NULL, OPEN_EXISTING, 0, NULL);
+	if (device == INVALID_HANDLE_VALUE) {
+		return false;
+	}
+
+	// Test to make sure this device is a CDROM drive
+	CDROM_TOC toc = {};
+	if (!DeviceIoControl(device, IOCTL_CDROM_READ_TOC, NULL, 0, &toc, sizeof(toc), NULL, NULL)) {
+		CloseHandle(device);
+		return false;
+	}
+
+	if (IsOpen()) {
+		CloseHandle(cdrom_handle);
+	}
+	cdrom_handle = device;
+	return true;
+}
+
+bool CDROM_Interface_Win32::SetDevice(const char* path)
+{
+	const char drive_letter = get_drive_letter(path);
+	if (!drive_letter) {
+		return false;
+	}
+	if (!Open(drive_letter)) {
+		return false;
+	}
+	InitAudio();
+	return true;
+}
+
+bool CDROM_Interface_Win32::GetUPC(unsigned char& attr, std::string& upc)
+{
+	if (!IsOpen()) {
+		return false;
+	}
+	CDROM_SUB_Q_DATA_FORMAT format = {};
+	format.Format = IOCTL_CDROM_MEDIA_CATALOG;
+	SUB_Q_CHANNEL_DATA data        = {};
+	if (!DeviceIoControl(cdrom_handle, IOCTL_CDROM_READ_Q_CHANNEL, &format, sizeof(format), &data, sizeof(data), NULL, NULL)) {
+		return false;
+	}
+	if (!data.MediaCatalog.Mcval) {
+		return false;
+	}
+	attr = 0;
+	upc = (const char *)data.MediaCatalog.MediaCatalog;
+	return true;
+}
+
+bool CDROM_Interface_Win32::GetAudioTracks(uint8_t& stTrack, uint8_t& end, TMSF& leadOut)
+{
+	if (!IsOpen()) {
+		return false;
+	}
+	CDROM_TOC toc = {};
+	if (!DeviceIoControl(cdrom_handle, IOCTL_CDROM_READ_TOC, NULL, 0, &toc, sizeof(toc), NULL, NULL)) {
+		return false;
+	}
+	if (toc.LastTrack >= MAXIMUM_NUMBER_TRACKS) {
+		return false;
+	}
+	stTrack = toc.FirstTrack;
+	end     = toc.LastTrack;
+	leadOut.min = toc.TrackData[toc.LastTrack].Address[1];
+	leadOut.sec = toc.TrackData[toc.LastTrack].Address[2];
+	leadOut.fr  = toc.TrackData[toc.LastTrack].Address[3];
+	return true;
+}
+
+bool CDROM_Interface_Win32::GetAudioTrackInfo(uint8_t track, TMSF& start, unsigned char& attr)
```

Comment:
const

Reply:
`track` could be. `start` and `attr` are output parameters.